openedx/core/djangoapps/course_groups/views.py

```diff
@@ -29,76 +31,119 @@ def json_http_response(data):
 
 def split_by_comma_and_whitespace(cstr):
     """
-    Split a string both by commas and whitespice.  Returns a list.
+    Split a string both by commas and whitespace.  Returns a list.
     """
     return re.split(r'[\s,]+', cstr)
 
 
-@ensure_csrf_cookie
-def list_cohorts(request, course_key_string):
+def link_cohort_to_partition_group(cohort, partition_id, group_id):
     """
-    Return json dump of dict:
-
-    {'success': True,
-     'cohorts': [{'name': name, 'id': id}, ...]}
+    Create cohort to partition_id/group_id link.
     """
+    CourseUserGroupPartitionGroup(
+        course_user_group=cohort,
+        partition_id=partition_id,
+        group_id=group_id,
+    ).save()
 
-    # this is a string when we get it here
-    course_key = SlashSeparatedCourseKey.from_deprecated_string(course_key_string)
 
-    course = get_course_with_access(request.user, 'staff', course_key)
+def unlink_cohort_partition_group(cohort):
+    """
+    Remove any existing cohort to partition_id/group_id link.
+    """
+    CourseUserGroupPartitionGroup.objects.filter(course_user_group=cohort).delete()
 
-    all_cohorts = [
-        {
-            'name': c.name,
-            'id': c.id,
-            'user_count': c.users.count(),
-            'assignment_type': cohorts.CohortAssignmentType.get(c, course)
-        }
-        for c in cohorts.get_course_cohorts(course)
-    ]
 
-    return json_http_response({'success': True,
-                               'cohorts': all_cohorts})
+def _get_cohort_representation(cohort, course):
+    """
+    Returns a JSON representation of a cohort.
+    """
+    group_id, partition_id = cohorts.get_group_info_for_cohort(cohort)
+    return {
+        'name': cohort.name,
+        'id': cohort.id,
+        'user_count': cohort.users.count(),
+        'assignment_type': cohorts.CohortAssignmentType.get(cohort, course),
+        'user_partition_id': partition_id,
+        'group_id': group_id
+    }
 
 
+@require_http_methods(("GET", "PUT", "POST", "PATCH"))
 @ensure_csrf_cookie
-@require_POST
-def add_cohort(request, course_key_string):
+@expect_json
+@login_required
+def cohort_handler(request, course_key_string, cohort_id=None):
     """
-    Return json of dict:
-    {'success': True,
-     'cohort': {'id': id,
-                'name': name}}
-
-                or
-
-    {'success': False,
-     'msg': error_msg} if there's an error
+    The restful handler for cohort requests. Requires JSON.
+    GET
+        If a cohort ID is specified, returns a JSON representation of the cohort
+            (name, id, user_count, assignment_type, user_partition_id, group_id).
+        If no cohort ID is specified, returns the JSON representation of all cohorts.
+           This is returned as a dict with the list of cohort information stored under the
+           key `cohorts`.
+    PUT or POST or PATCH
```

Comment:
I'm no REST expert, but I think POST makes sense in the case where a client POSTs to `courses/<course_key>/cohorts`. However, when creating a new cohort at `courses/<course_key>/cohorts/<new_cohort_id>`, PUT should be used.

Source: http://stackoverflow.com/questions/630453/put-vs-post-in-rest